### PR TITLE
[TensorIR][PASS][M1c] CompactBufferAllocation

### PR DESCRIPTION
**Pull request by Hzfengsy**
This PR is part of the TensorIR upstreaming effort (#7527) as one of the M1c stages, introducing the pass `ConvertBlocksToOpaque` and `CompactBufferAllocation`

Pass `ConvertBlocksToOpaque` will substitute all the block vars with the PrimExprs they are bound to its iter_values and convert blocks into opaque ones.

Pass `CompactBufferAllocation` compacts the buffer access region by removing the buffer regions that are not accessed.

These two passes both are important passes for converting schedulable TensorIR to low-level tir.

cc @tqchen @junrushao1994 @comaniac @xqdan @jroesch 
cc @tqchen @comaniac @jroesch @junrushao1994

### CODE REVIEW

**Review by tqchen — APPROVED** [review]

**Review by comaniac — APPROVED** [review]
LGTM. Just nits.

File: include/tvm/tir/stmt.h [hunk 1]
```diff
@@ -991,13 +992,22 @@ class BufferRegion : public ObjectRef {
...
    */
   TVM_DLL static BufferRegion FullRegion(Buffer buffer);
 
+  /*!
+   * \brief Create a BufferRegion which is a single point of the given buffer.
+   * \param buffer The buffer to generate single point BufferRegion.
+   * \param indices The access point indices of the buffer
+   * \return The BufferRegion which is the single point of the given buffer.
+   */
+  TVM_DLL static BufferRegion FromPoint(Buffer buffer, Array<PrimExpr> indices);
```
[the remark this answers is not in the document]
Comment (Hzfengsy) [reply include/tvm/tir/stmt.h:1007]: `FromRegion` is not accurate. I guess `FullRegion` is good or we can call it `FromFullRegion`

File: python/tvm/tir/transform/transform.py [hunk 2]
```diff
@@ -560,3 +560,53 @@ def PlanAndUpdateBufferAllocationLocation():
...
+
+    Returns
+    -------
+    fpass : tvm.transform.Pass
+        The result pass
+    """
+    return _ffi_api.ConvertBlocksToOpaque()
+
+
+def CompactBufferAllocation():
```
Comment (xqdan) [inline python/tvm/tir/transform/transform.py:578]: So，is this pass doing the same job as InferBound?

**Review by junrushao1994 — APPROVED** [review]

File: python/tvm/tir/transform/transform.py (hunk 2 above)
Comment (tqchen) [reply python/tvm/tir/transform/transform.py:578]: Yes, the functionality should be similar, except that infer bound needs to walk through the schedule tree, while in this case the split/reorder already updated the index, so the impl should be simpler

**Review by xqdan — APPROVED** [review]

File: python/tvm/tir/transform/transform.py (hunk 2 above)
Comment (xqdan) [reply python/tvm/tir/transform/transform.py:578]: Thanks for explaination!

**Review by jcf94 — APPROVED** [review]

**Comment by jcf94** [comment]
Merged. Thanks! @Hzfengsy @tqchen @comaniac @junrushao1994 @xqdan

**Pull request merged**